rs/sns/governance/src/governance.rs

```diff
@@ -2615,6 +2691,63 @@ impl Governance {
         Ok(false)
     }
 
+    async fn upgrade_sns_framework_canister(
+        &mut self,
+        new_wasm_hash: Vec<u8>,
+        canister_type_to_upgrade: SnsCanisterType,
+    ) -> Result<bool, GovernanceError> {
+        let root_canister_id = self.proto.root_canister_id_or_panic();
+
+        let target_wasm = get_wasm(&*self.env, new_wasm_hash.to_vec(), canister_type_to_upgrade)
+            .await
+            .map_err(|e| {
+                GovernanceError::new_with_message(
+                    ErrorType::External,
+                    format!("Could not get wasm for upgrade: {}", e),
+                )
+            })?
+            .wasm;
+
+        let target_is_root = canister_type_to_upgrade == SnsCanisterType::Root;
+
+        if target_is_root {
+            upgrade_canister_directly(
+                &*self.env,
+                root_canister_id,
+                target_wasm,
+                Encode!().unwrap(),
+            )
+            .await?;
+        } else {
+            let canister_ids_to_upgrade =
+                get_canisters_to_upgrade(&*self.env, root_canister_id, canister_type_to_upgrade)
+                    .await
+                    .map_err(|e| {
+                        GovernanceError::new_with_message(
+                            ErrorType::External,
+                            format!("Could not get list of SNS canisters from root: {}", e),
+                        )
+                    })?;
+            for target_canister_id in canister_ids_to_upgrade {
+                self.upgrade_non_root_canister(
```

Comment:
Just so you know, this is going to return before the upgrade is actually finished, which I don't think works correctly with the control flow later on.